File: haskell/grains/src/Grains.hs
module Grains
  ( square
  , total
  ) where

square :: Integer -> Maybe Integer
square n
  | n >= 1 && n <= 64 = Just (2 ^ (n - 1))
  | otherwise = Nothing

total :: Integer
total = 2 ^ (64 :: Int) - 1

File: cpp/grains/grains.cpp
#include "grains.h"

namespace grains
{
unsigned long long square(int n)
{
    return 1ULL << (n - 1);
}

unsigned long long total()
{
    return (1ULL << 63) - 1 + (1ULL << 63);
}
} // namespace grains
